Guide for Flax partitioning (auto SPMD) API

IvyZX opened the pull request:
Original code PR: https://github.com/google/flax/pull/2704

marcvanzee commented:
@IvyZX the doc build is failing, some problem with the sharding code. Could you please fix it?

levskaya commented on docs/guides/spmd_api.ipynb:
```diff
@@ -0,0 +1,1135 @@
+{
```
<p>"scaled module output"  what does "scaled" mean here? do you mean "sharded" or am I missing something?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/google/flax/pull/2730/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='17'/>

IvyZX replied:
Removed "scaled" - hopefully it's clearer now.

levskaya commented on docs/guides/spmd_api.ipynb:
```diff
@@ -0,0 +1,1135 @@
+{
```
<p>"Profilling" --&gt; "Profiling"  (one "L")</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/google/flax/pull/2730/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='26'/>

IvyZX replied:
Done.

levskaya commented on docs/guides/spmd_api.ipynb:
```diff
@@ -0,0 +1,1135 @@
+{
```
<p>"replaced"-&gt;"replaces"</p><p> </p><p>We should perhaps note that there are more complicated/advanced sharding patterns that require annotating -activation- dimension names differently from -parameter- dimension names.  For really advanced stuff, people may wish to do fine-grained manual mesh assignments using the simpler system above, but the logical-naming helpers are useful for getting started in exploring different sharding layouts.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/google/flax/pull/2730/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='28'/>

IvyZX replied:
Added a new section at the bottom of logic axis introduction. Let me know if there could be a better phrasing of the idea!

IvyZX commented:
> @IvyZX the doc build is failing, some problem with the sharding code. Could you please fix it?

This seems to be due to JAX dropping Python 3.7. I switched readthedocs build to 3.8 and the test passes now.

marcvanzee commented on docs/guides/index.rst:
```diff
@@ -14,6 +14,7 @@ Guides
    extracting_intermediates
    lr_schedule
    use_checkpointing
+   spmd_api
```
Maybe this is an advanced guide? It looks quite involved and you do a number of advanced things (lifted transformations, scan over layers, ...)

IvyZX replied:
Lifted transformations is the only advanced topic here, and it is not really explored in detail. Since JAX team may want to upstream `pjit` to `jit` in the future, it may make sense to put it on the later section of guides as an introduction on how to scale up a model. (JAX also put their [pjit doc](https://jax.readthedocs.io/en/latest/jax-101/08-pjit.html) as the last piece of their 101 tutorial.)

IvyZX commented on docs/guides/spmd_api.md:
```diff
@@ -0,0 +1,432 @@
+
+    z = jnp.dot(y, W2)
+    # force a local sharding annotation, fairly redundant in this case.
+    z = with_sharding_constraint(z, PartitionSpec('data', None))
```
Because the output should be sharded as ('data', None), in which 'data' refers to the batch dimension of input.

IvyZX commented on docs/guides/spmd_api.md:
```diff
@@ -0,0 +1,432 @@
+        (x.shape[-1], self.depth))
+
+    y = jax.nn.relu(jnp.dot(x, W1))
+    # force a local sharding annotation, fairly redundant in this case.
```
The local sharding annotation here. I removed the comment here and added more explanation on the documentation above. Basically `with_sharding_constraint` is only to ensure that `pjit` shards this variable in this way, and it's redundant here because `pjit` will by default shard it this way too.

marcvanzee replied:
Ah makes sense, thanks for updating it!

IvyZX commented on docs/guides/spmd_api.md:
```diff
@@ -0,0 +1,432 @@
+
+
+
+Each value passed into a sharding annotates a dimension, and its value should be among `data`, `model` or `None`. This refers to whether the data's axis should be sharded across one of the device mesh dimensions, or not sharded at all.
```
Yes, this is correct. I rephrased it to:

Calling these APIs annotates each dimension of your parameter and intermediate variables with an axis name, namely `'data'`, `'model'` or `None`. This refers to how this dimension should be sharded - across one of the device mesh dimensions, or not sharded at all.

For example, if we define `W1`, which has shape `(x.shape[-1], self.depth)`, with `(None, 'model')`, the first dimension (of length `x.shape[-1]`) will be replicated across all devices, while the second dimension (of length `self.depth`) will be sharded over the `model` axis of the device mesh.

IvyZX commented:
Thank you Marc! I address all your comments, either fixing them directly or leaving some comments beneath yours. Please take a look and let me know what you think.

8bitmp3 commented on docs/guides/spmd_api.md:
```diff
@@ -0,0 +1,429 @@
+---
+
+<!-- #region id="2-JYevEklHfB" -->
+# Scale up Flax modules with `pjit`
```
Suggestion:

Maybe mention multi-host/multi-devices in the title so that users know what this scaling up with `pjit` means without indepth JAX knowledge. Similar to "Using JAX in multi-host and multi-process environments" https://jax.readthedocs.io/en/latest/multi_process.html

8bitmp3 commented on docs/guides/spmd_api.md:
```diff
@@ -0,0 +1,429 @@
+
+In this guide, you will learn how to use JAX's `pjit` and `flax.linen.spmd` to scale up your Flax modules on multiple devices and hosts.
+
+Modern JAX provides [`jax.experimental.pjit`](https://jax.readthedocs.io/en/latest/jax.experimental.pjit.html) as a way to automatically compile and scale up JAX computations using the [SPMD](https://jax.readthedocs.io/en/latest/glossary.html?highlight=spmd#term-SPMD) model. You can think of `pjit` as [`jit`](https://jax.readthedocs.io/en/latest/jax-101/02-jitting.html) where one explicitly speci ...
```
Suggestions:

- Do we need "modern"? (Currently, there's just one JAX 0.4, isn't it?)
- Spell out "Single Program Multi Data" before abbreviating it.
- Use the SPMD link without the `highlight=...` part like `https://jax.readthedocs.io/en/latest/glossary.html#term-SPMD)`
- Last sentence:
  - "You can think of `pjit` as [`jax.jit`]..." (with `jax.jit`).
  - Use "Refer to" instead of "See" (more accessibility-friendly language), "JAX guide" -> "JAX-101 `pjit` tutorial" (more precise).
  - `pjit` (formatting). 

```suggestion
JAX provides [`jax.experimental.pjit`](https://jax.readthedocs.io/en/latest/jax.experimental.pjit.html) as a way to automatically compile and scale up JAX computations using the [Single Program Multiple Device (SPMD)](https://jax.readthedocs.io/en/latest/glossary.html#term-SPMD) model. You can think of JAX's `pjit` as [`jax.jit`](https://jax.readthedocs.io/en/latest/jax-101/02-jitting.html) where you can explicitly specify how the input and output data should be partitioned across devices. Refer to the [JAX-101 `pjit` tutorial](https://jax.readthedocs.io/en/latest/jax-101/08-pjit.html) for more information on `pjit`.
```

8bitmp3 reviewed (commented):
@IvyZX A few suggestions for the first part of the guide. Thanks for putting this together! Amazing doc

marcvanzee commented on docs/guides/spmd_api.md:
```diff
@@ -0,0 +1,429 @@
+<!-- #region id="XfuDTqt1mO6g" -->
+Import all the `pjit` related libraries. Note that they are still in the experimental package of JAX, so some API may change. In the near future, the plan is to move the `pjit` API out of the experimental status and merge it with the current `jax.jit` API.
+
+Start a device mesh using the 8 devices available. In this example, we set them as a `2x4` device mesh, same as the layout of TPU v3-8, and annotate each axis with a name. A typical way to annotate axis names is `('data', 'model')`: `'data'` being the mesh dimension used for data-parallel sharding of the batch dimension of inputs and activations, and the `'model'` being the mesh dimension used fo ...
```
Thinking about this some more, if you have a `2x4` device and you annotate axis names with `('data', 'model')`, then this mean that if you partition over the `data` dimension you are effectively doing a two-way partition over the first dimension of the devices, and if you partition over the `model` dimension you partition it four way over the second dimension, right? It might be good to explain this explicitly, so the relation between the devices and the axis names becomes clearer.

marcvanzee commented on docs/guides/spmd_api.md:
````diff
@@ -0,0 +1,429 @@
+```
+
+<!-- #region id="3d6b15d08980" -->
+However, to generate `PartitionSpec` for the output, we need to use some actual output as reference. One solution is to create a model and evaluate `model.init` abstractly using `jax.eval_shape`, and then use `nn.get_partition_spec` to automatically generate the `PartitionSpec`. 
````
"we need to use some actual output as reference" --> don't we need to generate the variable dict structure rather than the model outputs?

IvyZX replied:
That works too. I am just trying to show a more complicated example here, in which the pjitted function output is another dataclass instead of the simple variable dict. Added the explanation in doc.

marcvanzee commented on docs/guides/spmd_api.md:
```diff
@@ -0,0 +1,429 @@
+
+> A side note: Here we define our `init_fn` as purely functional and takes `model` and `optimizer` as arguments. This is not necessary - you can simply define with `def init_fn(k, x):` and all will work fine here.
+> 
+> This guide doesn't do it because later we will show you another way to define your model and will run `init_fn` with another model instance. But this does a bit of trouble because `jax.eval_shape` only takes numeric inputs, and we have to create an abstract closure before feeding the function in.
```
```suggestion
> This guide doesn't do it because later we will show you another way to define your model and will run `init_fn` with another model instance. However, this is problematic because `jax.eval_shape` only takes numeric inputs, so we have to create an abstract closure before feeding the function in.
```

marcvanzee commented on docs/guides/spmd_api.md:
```diff
@@ -0,0 +1,429 @@
+<!-- #region id="1gvf6_a_OeU3" -->
+Now define the `MLP` model as multiple layers of `DotReluDot`.
+
+Note that you can use `nn.scan` and other [lifted transforms](https://flax.readthedocs.io/en/latest/advanced_topics/lift.html) on the module. This will introduce another dimension on the params, and this dimension should be annotated in `metadata_params`. Since the parameters inside `DotReluDot` are already sharded along `model` axis, we don't need to partition multiple layers across `model` dime ...
```
```suggestion
Note that you can use `nn.scan` and other [lifted transforms](https://flax.readthedocs.io/en/latest/advanced_topics/lift.html) on the module. This will introduce another dimension for the params (the dimension over which `scan` is applied), and this dimension should be speficied using the `metadata_params` argument. Since the parameters inside `DotReluDot` are already sharded along `model` axis, we don't need to partition multiple layers across `model` dimension here, and therefore denote it as `None`.
```

marcvanzee commented on docs/guides/spmd_api.md:
````diff
@@ -0,0 +1,429 @@
+# we need a rule on every leaf of the input variables, in this case a nested dict.
+state_spec = nn.get_partition_spec(abstract_variables)
+state_spec
+```
````
The output shows now three dimensions for the `PartitionSpec`, while you only provide 2 as input. I suppose this is because of the scan, but it is a but surprising so it might be worth clarifying this.

IvyZX replied:
The PartitionSpec shows that the params W1 and W2 has three dimensions, which WAI. The output of both pjitted functions are `TrainState`, which contains the params.

marcvanzee commented on docs/guides/spmd_api.md:
````diff
@@ -0,0 +1,429 @@
+```
+
+<!-- #region id="KAyPfNP9pTR4" -->
+# `pjit` the initialization and train step
````
Initialization with pjit is complex! Quite a number of steps, maybe we can simplify this with a simple utility function (not now, but maybe as a follow-up?)

IvyZX replied:
You can use decorator `@functools.partial(pjit, ...)` like `jit`. I added the explanation in the doc.

marcvanzee commented on docs/guides/spmd_api.md:
````diff
@@ -0,0 +1,429 @@
+```
+
+<!-- #region id="dcLFwH1dhr3v" -->
+## `pjit` the train step and inference 
````
It isn't clear which parts of these code blocks refer to "train step" and which refer to "inference". Maybe add some text before the inference code block to clarify that.

marcvanzee commented on docs/guides/spmd_api.md:
```diff
@@ -0,0 +1,429 @@
+<!-- #region id="vXY2mx7WjOyv" -->
+## Profiling
+
+If you are running on a TPU pod or pod slice, you can use the `block_all` util below to profile the cell below to observe a performance increase.
```
```suggestion
If you are running on a TPU pod or pod slice, you can use the `block_all` utility function below to observe a performance increase.
```

andsteing replied:
how do I see the "performance increase" ?

marcvanzee commented on docs/guides/spmd_api.md:
```diff
@@ -0,0 +1,429 @@
+<!-- #region id="GUXkJiQkotfm" -->
+# Logical axis annotation
+
+JAX auto SPMD encourages user to explore different sharding layouts to find the optimal one. To this end, in Flax you actually can annotate the dimensions of any array with more descriptive axis names, instead of only the device mesh axis names (i.e., `data` and `model`). 
```
```suggestion
JAX auto SPMD encourages users to explore different sharding layouts to find the optimal one. To this end, in Flax you actually can annotate the dimensions of any array with more descriptive axis names, instead of only the device mesh axis names (i.e., `data` and `model`). 
```

marcvanzee commented on docs/guides/spmd_api.md:
```diff
@@ -0,0 +1,429 @@
+
+Check out the `Logical-` model definitions below. It's exactly the same with the model above, except for two differences:
+
+1. Each axis are annotated with more concrete, meaningful names - like `embed`, `hidden`, `batch` and `layer`. These names are referred as "logical axis names" in Flax. They make the dimensional changes inside model definition more readable.
```
```suggestion
1. All axes are annotated with more concrete, meaningful names - like `embed`, `hidden`, `batch` and `layer`. These names are referred as "logical axis names" in Flax. They make the dimensional changes inside model definitions more readable.
```

marcvanzee reviewed (approved):
Awesome tutorial @IvyZX, great work and thanks a lot! Just some minor comments but overall LGTM.

andsteing commented on docs/guides/spmd_api.md:
````diff
@@ -0,0 +1,429 @@
+```
+
+```python id="F5gy_gb_mOqR"
+with mesh:
````
nit: in the other examples, the `with mesh:` code is in the same cell as the function being pjitted

andsteing reviewed (commented):
Great guide! This is very helpful to make use of `nn.with_partitioning` and friends.

I'm fairly new to `pjit`, so I also marked some things where I had genuine understanding questions.

cgarciae commented:
Thanks Ivy! This material is super useful, I've been reading about this so personally its great to have this. Just a few remarks:
1. Given that JAX 0.4 is out, should we consider using the new APIs (e.g. `device_put` + `jit`)? I am a bit biased because this is how I am learning.
2. On one hand, showing the `scan` trick is nice as its used in real life, on the other hand its and additional complexity that might not be ideal as its yet another concept in an already packed guide.
3. Some references or explanation as to why the parameters in `DotReluDot` are sharded the way there are would be nice. I just happened learn about this before reading this guide so it made perfect sense but without it I'd have some questions.

marcvanzee commented:
> On one hand, showing the scan trick is nice as its used in real life, on the other hand its an additional complexity that might not be ideal as its yet another concept in an already packed guide.

+1, I think scan over layers deserves it own guide and right now it is a bit much to take in here.

But to avoid stalling this guide, maybe we can keep it in but @IvyZX  can file an issue saying we want to add a guide for scan over layers, and then link to it from here?

cgarciae commented:
@marcvanzee agreed. Lets merge this PR and improve the guide in the near future. Maybe just collect some of the points in this comments in an issue.

IvyZX commented on docs/guides/spmd_api.md:
```diff
@@ -0,0 +1,429 @@
+---
+
+<!-- #region id="2-JYevEklHfB" -->
+# Scale up Flax modules with `pjit`
```
Renamed to `flax_on_pjit.md`.

andsteing replied:
`flax_on_pjit.md` is better, but it still does not really match the title `Scale up Flax modules on multiple devices` ...

IvyZX commented on docs/guides/spmd_api.md:
```diff
@@ -0,0 +1,429 @@
+
+In this guide, you will learn how to use JAX's `pjit` and `flax.linen.spmd` to scale up your Flax modules on multiple devices and hosts.
+
+Modern JAX provides [`jax.experimental.pjit`](https://jax.readthedocs.io/en/latest/jax.experimental.pjit.html) as a way to automatically compile and scale up JAX computations using the [SPMD](https://jax.readthedocs.io/en/latest/glossary.html?highlight=spmd#term-SPMD) model. You can think of `pjit` as [`jit`](https://jax.readthedocs.io/en/latest/jax-101/02-jitting.html) where one explicitly speci ...
```
Interesting. I rewrote the introductory paragraph with more basic explanation.

IvyZX commented on docs/guides/spmd_api.md:
````diff
@@ -0,0 +1,429 @@
+<!-- #endregion -->
+
+```python id="e3d005e672d6"
+# TODO: replace this with `pip3 install flax` when 0.6.4 release rolls out.
````
Since this is a public doc, it's probably better to not expose our names. I can open an issue to remind myself of that.

8bitmp3 replied:
fyi agreed with @IvyZX 

```
# Once Flax v0.6.4 is out, replace this with `pip3 install flax`.
```

This way, the docs can be more or less up-to-date after 0.6.4.

andsteing replied:
You could use Github username (which is already known) ... Or create an issue and assign it to yourself.

The main motivation here is to avoid accumulating TODOs in public docs.

IvyZX commented on docs/guides/spmd_api.md:
````diff
@@ -0,0 +1,429 @@
+```
+
+<!-- #region id="XfuDTqt1mO6g" -->
+Import all the `pjit` related libraries. Note that they are still in the experimental package of JAX, so some API may change. In the near future, the plan is to move the `pjit` API out of the experimental status and merge it with the current `jax.jit` API.
````
Moving extra comments in note.

IvyZX commented on docs/guides/spmd_api.md:
```diff
@@ -0,0 +1,429 @@
+<!-- #region id="1gvf6_a_OeU3" -->
+Now define the `MLP` model as multiple layers of `DotReluDot`.
+
+Note that you can use `nn.scan` and other [lifted transforms](https://flax.readthedocs.io/en/latest/advanced_topics/lift.html) on the module. This will introduce another dimension on the params, and this dimension should be annotated in `metadata_params`. Since the parameters inside `DotReluDot` are already sharded along `model` axis, we don't need to partition multiple layers across `model` dime ...
```
I put some more concrete explanation in `metadata_params`.

IvyZX commented on docs/guides/spmd_api.md:
```diff
@@ -0,0 +1,429 @@
+      x, _ = nn.scan(DotReluDot, length=self.num_layers, 
+                    variable_axes={"params": 0},
+                    split_rngs={"params": True},
+                    metadata_params={nn.PARTITION_NAME: None}
```
This is just a constant to denote the partition name that annotates the `nn.scan` layer.

andsteing replied:
it's clearer now with the added "Note:"

IvyZX commented on docs/guides/spmd_api.md:
```diff
@@ -0,0 +1,429 @@
+
+Note that in the output `initialized_state`, the params `W1` and `W2` are of type [`Partitioned`](https://flax.readthedocs.io/en/latest/api_reference/_autosummary/flax.linen.Partitioned.html). This is a wrapper around the actual JAX array that allows Flax to record metadata associated with it. You can access the raw JAX array by adding `.value` or running `.unbox()`.
+
+You can also check the underlying `.sharding` of the JAX array, which gives a hint on the way it is partitioned.
```
This is a fairly new concept in JAX and is not very well documented as for now. I only find one mentioning of OpSharding here: https://jax.readthedocs.io/en/latest/jax_array_migration.html#why-create-jax-array

IvyZX commented on docs/guides/spmd_api.md:
```diff
@@ -0,0 +1,429 @@
+pjit_init_fn = pjit(init_fn,
+                    static_argnums=(2, 3),
+                    in_axis_resources=(PartitionSpec(None), x_spec),  # RNG key and x
+                    out_axis_resources=state_spec
```
JAX team is making `pjit` more automatic so that user don't have to specify input and output axis. But as far as Anselm and I know, this approach is not very efficient on complex model for now, so we decided to still present the version that explicitly states the axis.

andsteing replied:
ah ... so it's still a bit rough around the edges, I see.

I think I would still mention this because in many examples users will see examples of `pjit()` where the `out_axis_resources` is omitted (e.g. our very own `flax.core.meta.Partitioned`).

IvyZX commented:
Thank you all for the reviews! I have addressed all the comments listed here, either by a reply comment or by directly modifying the doc content. PTAL and let me know if there's any questions/concerns.

andsteing commented on docs/guides/flax_on_pjit.md:
```diff
@@ -0,0 +1,437 @@
+
+To replicate identical layers, you can either use `flax.linen.scan` or a for-loop: `flax.linen.scan` can offer faster compilation times, whereas the for-loop can be faster on runtime. This guide uses `flax.linen.scan` simply to show that [Flax lifted transforms](https://flax.readthedocs.io/en/latest/advanced_topics/lift.html#supported-transformations) work together with [JAX `pjit`](https://jax.r ...
+
+**Note:** `flax.linen.scan` will introduce another dimension for the params (the dimension over which `scan` is applied), and you need to use the `metadata_params` argument to annotate the partition of this dimension. Since the parameters inside your `DotReluDot` (a sub-`Module`) are already sharded along the `model` axis, you don't need to partition multiple layers across the `model` dimension h ...
```
this "note" is formatted differently from other "notes" - is this on purpose?
![image](https://user-images.githubusercontent.com/16050001/209135442-fb5202c5-26ab-4cff-ae82-1b2d7b99e63c.png)
vs
![image](https://user-images.githubusercontent.com/16050001/209135488-7e447e95-5e56-4154-a772-8b3e0fc91518.png)

8bitmp3 replied:
@andsteing We are doing a 2nd (3rd?) review with @IvyZX asap, going through the second half. Will make sure "Note:" is formatted consistently :+1: Thanks for the feedback!!

andsteing reviewed (approved):
I had another go at the updated guide. Nice improvements, thanks!

I left some more comments, but will approve here, since it's mostly nits and nothing is blocking.

8bitmp3 commented:
Last review done https://github.com/google/flax/pull/2730/commits/dcf5517e662f925f28327755fbdd225741e99ac4. Thanks @IvyZX for the amazing work :+1:
